server/src/main/java/org/elasticsearch/common/logging/DeprecationIndexingService.java

```diff
@@ -0,0 +1,154 @@
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.client.OriginSettingClient;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.metadata.ComposableIndexTemplate;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.xcontent.NamedXContentRegistry;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.json.JsonXContent;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.elasticsearch.common.Strings.isNullOrEmpty;
+
+/**
+ * This service is responsible for writing deprecation messages to a data stream. It also creates
+ * the data stream if necessary. The writing of messages can be toggled using the
+ * {@link #WRITE_DEPRECATION_LOGS_TO_INDEX} setting.
+ */
+public class DeprecationIndexingService implements ClusterStateListener {
+    private static final Logger LOGGER = LogManager.getLogger(DeprecationIndexingService.class);
+
+    private static final String DATA_STREAM_NAME = "logs-deprecation-elasticsearch";
+    private static final String TEMPLATE_NAME = DATA_STREAM_NAME + "-template";
+    private static final String TEMPLATE_MAPPING = TEMPLATE_NAME + ".json";
+
+    private static final String DEPRECATION_ORIGIN = "deprecation";
+
+    public static final Setting<Boolean> WRITE_DEPRECATION_LOGS_TO_INDEX = Setting.boolSetting(
+        "cluster.deprecation_indexing.enabled",
+        true,
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
+    );
+
+    private final Client client;
+    private boolean hasTriedToLoadTemplate = false;
+    private volatile boolean isEnabled = false;
+
+    public DeprecationIndexingService(ClusterService clusterService, Client client) {
+        this.client = new OriginSettingClient(client, DEPRECATION_ORIGIN);
+
+        clusterService.addListener(this);
+    }
+
+    /**
+     * Indexes a deprecation message.
+     * @param key       the key that was used to determine if this deprecation should have been be logged.
+     *                  Useful when aggregating the recorded messages.
+     * @param message   the message to log
+     * @param xOpaqueId the associated "X-Opaque-ID" header value if any, or <code>null</code>
+     * @param params    parameters to the message, if any
+     */
+    public void writeMessage(String key, String message, String xOpaqueId, Object[] params) {
+        if (this.isEnabled == false) {
+            return;
+        }
+
+        Map<String, Object> payload = new HashMap<>();
+        payload.put("@timestamp", Instant.now().toString());
+        payload.put("key", key);
+        payload.put("message", message);
```

Comment:
message and params I think should be 'expanded' otherwise `deprecationLogger.deprecate("msg {}", 123) ` would create two fields `"message: " msg" , "params" : 123` instead of just `msg 123`
or we could think of using log4j layout as per previous comment

Reply:
I actually did that intentionally, because I thought it might be more useful...for some reason. Possibly that was before I changed the code to always require a key, so maybe now it makes sense to expand the parameters?

Reply:
I think it would be better to expand them before indexing. 
Re `key` field I don't think we need it. We do not log `key` in deprecation json (nor plaintext before) .
btw Would it be far more expensive to index every deprecation log message (no rate limiting)?